DAOS-14804 vos: fix memory track counter

wangshilong opened the pull request:
1. remove duplicated vtl_dtx_cmt_ent_cnt.
2. mallocinfo() hurts performance significantly, only got about 100MiB per seconds
output with memory track enabled, reduce frequency of recording.
3. fix per xstream memory track counter.


Required-githooks: true

Commit:
DAOS-14804 vos: fix memory track counter

1. @vtl_committed metric created twice
2. @vtl_dtx_cmt_ent_cnt should be not increased if there are some errors,
also it should be decreased properly in some cases like container destroy.

Required-githooks: true
Signed-off-by: Wang Shilong <shilong.wang@intel.com>

Nasf-Fan commented on src/vos/vos_container.c:
```diff
@@ -200,6 +200,8 @@ cont_free_internal(struct vos_container *cont)
 	d_tm_dec_gauge(vos_tls_get(cont->vc_pool->vp_sysdb)->vtl_committed,
 		       cont->vc_dtx_committed_count);
+	d_tm_dec_gauge(vos_tls_get(cont->vc_pool->vp_sysdb)->vtl_dtx_cmt_ent_cnt,
+		       cont->vc_dtx_committed_count);
```
I am wondering why we maintain both "vtl_committed" and "vtl_dtx_cmt_ent_cnt"? They seems have the same functionality for the count of committed DTX entries. @wangdi1

wangdi1 replied:
ah, if they are the same, then maybe just keep one here. thanks

wangshilong replied:
Removed.

Commit:
more fixes

1. remove duplicated vtl_dtx_cmt_ent_cnt.
2. mallocinfo() hurts performance significantly, only got about 100MiB per seconds
output with memory track enabled, heap is shared between different threads,
no need to call it per XS, and it hurs performance significantly,
only record on the first main XS instead.

Required-githooks: true
Signed-off-by: Wang Shilong <shilong.wang@intel.com>

daosbuild1 commented:
Test stage Functional on EL 8.8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13549/2/testReport/

jolivier23 reviewed (dismissed):
You'll need to adjust the description

Commit:
Merge branch 'master' of github.com:daos-stack/daos into shilongw/DAOS-14804
Required-githooks: true

daosbuild1 commented:
Test stage Functional on EL 8.8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13549/3/testReport/

Commit:
update test

Required-githooks: true

Signed-off-by: Wang Shilong <shilong.wang@intel.com>

daosbuild1 commented:
Test stage Functional Hardware Large completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13549/4/execution/node/1410/log

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13549/4/execution/node/1318/log

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13549/4/execution/node/1364/log

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13549/6/execution/node/417/log

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13549/5/execution/node/438/log

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13549/6/execution/node/438/log

Commit:
Merge branch 'master' of github.com:daos-stack/daos into shilongw/DAOS-14804

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13549/7/testReport/

daosbuild1 commented:
Test stage Functional Hardware Large completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13549/7/execution/node/1479/log

Commit:
Merge branch 'master' of github.com:daos-stack/daos into shilongw/DAOS-14804

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13549/8/testReport/

Commit:
fix per xstream counter

Required-githooks: true

Signed-off-by: Wang Shilong <shilong.wang@intel.com>

Commit:
Merge branch 'master' of github.com:daos-stack/daos into shilongw/DAOS-14804

Commit:
Merge branch 'master' of github.com:daos-stack/daos into shilongw/DAOS-14804

Commit:
revert ms_mallinfo to per xs

Required-githooks: true

Signed-off-by: Wang Shilong <shilong.wang@intel.com>

Commit:
make counter signed

Required-githooks: true
Signed-off-by: Wang Shilong <shilong.wang@intel.com>

Commit:
Merge branch 'master' of github.com:daos-stack/daos into shilongw/DAOS-14804
Required-githooks: true

daosbuild1 commented:
Test stage Functional on EL 8.8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13549/11/testReport/

kjacque commented on src/gurt/telemetry.c:
```diff
@@ -1814,6 +1846,31 @@ d_tm_record_meminfo(struct d_tm_node_t *metric)
+ * \param[in]	metric	Pointer to the metric
+ * \param[in]	bytes	number to be updated
+ */
+void
+d_tm_update_memusage(struct d_tm_node_t *metric, int bytes)
+{
+
+	if (metric == NULL)
+		return;
+
+	if (metric->dtn_type != D_TM_MEMUSAGE) {
+		D_ERROR("Failed to update memusage on item %s not a "
+			"memusage.  Operation mismatch: " DF_RC "\n",
+			metric->dtn_name, DP_RC(-DER_OP_NOT_PERMITTED));
+		return;
+	}
+	d_tm_node_lock(metric);
+	metric->dtn_metric->dtm_data.memusage += bytes;
+	d_tm_node_unlock(metric);
+
+}
```
Unclear to me why this new type of metric is needed. Isn't this still basically a gauge?

wangshilong replied:
This new counter metric might be negative, and current infra for gauge based on uint64_t. I think my use case is a bit special, so i add a new type here.

mjmac replied:
Why would memory usage be a negative value?

wangshilong replied:
Memory might be allocated and freed by different XS, using an atomic for all XS might be too much overhead. since this is almost a debug operation for developers, sum all all XS memory usage here is accurate and helpful.

NiuYawei replied:
It's bit weird to define a signed integer as "D_TM_MEMUSAGE", also the "int bytes" can only support up to 2GB bytes.
Why don't we just change the current gauge to signed 64bits integer? Or introduce a new "signed gauge"?

wangshilong replied:
memusage is defined as "long long int",  i thought arg "int bytes" might be fine for normal allocation, but i might declare it to "long long int" as well, I am not aware of any other case that might use signed gauge.....

just change the current gauge to signed 64 bits integer will break current API, eg: d_tm_get_gauge()  which is a public api 
@kjacque might confirm. another way is introduce new API like you said "signed gauge", i wondered if it really deserved to do so?

kjacque replied:
I don't think it is a good idea to change the type for gauge, as we have a number of consumers of metrics. Changing that would have definite ripple effects. If we must have a new type, I think signed gauge is not a bad idea.

However, I still do not totally buy into the idea that we need to account for negative values in this case. Are we actually seeing underflows of these metrics in the field, @wangshilong? If so, could we update the d_tm_dec_gauge function to check if input > current value, and if so, set the result to 0? Would that be adequate for accuracy if this is a debug metric we are OK with approximating?

NiuYawei reviewed (changes requested):
Overlooked the metric type changes in prior review.
